### DISCUSSION TITLE
Interest in adding ONNX support ?

### DISCUSSION THREAD

**Discussion by zstreet87** [lululxvi/deepxde#746] [Ideas]
I'm interesting in exporting a deepxde model to ONNX.  This is pretty straightforward using some of the backends.  For example, using PyTorch is as simple as `torch.onnx.export(model, "model_saved_file.onnx",...)`.  That's the gist of the command but needs model `inputs` and other arguments for protobuf (please see documentation).

https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html  

Since DeepXDE supports multiple backends, I figured having the functionality to transfer to one backend to another would be useful in general (for example, performance analysis with different backends).  Looking at DeepXDE briefly, adding the ability to talk to ONNX on the DeepXDE layer on top of the backends doesn't seem difficult.

Any interest?

**Comment by lululxvi** [comment]
Yes, good idea. Here is the code of saving:

https://github.com/lululxvi/deepxde/blob/532ac4c2d1efc24d14b3c50434d7c4949348c900/deepxde/model.py#L815

and restore:

https://github.com/lululxvi/deepxde/blob/532ac4c2d1efc24d14b3c50434d7c4949348c900/deepxde/model.py#L870

Feel free to submit a PR.

**Comment by zstreet87** [comment]
Okay great! I've been looking at the code and I am wondering what the best way to incorporate ONNX into your design.  

Each backend has a different interface to ONNX.  For example, PyTorch exports an ONNX model like `torch.onnx.export(model, args...)` and exporting a TensorFlow model to ONNX is different.  To stay in line with your design, I thought an ONNX class in `models.py` that is an interface to the supported ONNX backends.  This would require each supported backend has the implementation code.  Any thoughts and/or suggestions would be great!

**Comment by lululxvi** [comment]
I am not sure what the best way is, as I am not familiar with ONNX. What you describe seems good, i.e., we can implement the code for each backend separately.
